pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -115,18 +130,62 @@ func (t *Terminator) Evict(pods []*v1.Pod) {
 			}
 		}
 	}
-	// 2. Evict in order:
-	// a. non-critical non-daemonsets
-	// b. non-critical daemonsets
-	// c. critical non-daemonsets
-	// d. critical daemonsets
-	if len(nonCriticalNonDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalNonDaemon...)
-	} else if len(nonCriticalDaemon) != 0 {
-		t.evictionQueue.Add(nonCriticalDaemon...)
-	} else if len(criticalNonDaemon) != 0 {
-		t.evictionQueue.Add(criticalNonDaemon...)
-	} else if len(criticalDaemon) != 0 {
-		t.evictionQueue.Add(criticalDaemon...)
+
+	// EvictInOrder evicts only the first list of pods which is not empty
+	// future Evict calls will catch later lists of pods that were not initially evicted
+	t.EvictInOrder(
+		nonCriticalNonDaemon,
+		nonCriticalDaemon,
+		criticalNonDaemon,
+		criticalDaemon,
+	)
+}
+
+func (t *Terminator) EvictInOrder(pods ...[]*v1.Pod) {
+	for _, podList := range pods {
+		if len(podList) > 0 {
+			// evict the first list of pods that is not empty, ignore the rest
+			t.evictionQueue.Add(podList...)
+			return
+		}
 	}
 }
+
+func (t *Terminator) DeleteExpiringPods(ctx context.Context, pods []*v1.Pod, nodeGracePeriodTerminationTime *time.Time) error {
+	for _, pod := range pods {
+		// check if the node has an expiration time and the pod needs to be deleted
+		deleteTime := t.podDeleteTimeWithGracePeriod(nodeGracePeriodTerminationTime, pod)
+		if deleteTime != nil && time.Now().After(*deleteTime) {
```

Comment:
I think I might've left this comment before but it got lost. Is there a reason why we're not doing the IsZero() check? 
```suggestion
		if !deleteTime.IsZero() && time.Now().After(*deleteTime) {
```

Reply:
I took the prior suggestion to use `IsZero()`, I think this is another case. I'll look for any more that might have been missed.

Reply:
I reverted these back to a nil check to avoid juggling metav1.Time vs time.Time everywhere. We're managing an annotation rather than a formal field on a CRD, so I think the simplicity we gain from avoiding metav1.Time is worth it.

If a user opts to set this field themselves for some reason and the timestamp fails to parse, the behavior falls back as if the terminationGracePeriod were not set for the NodePool. @jmdeal also agrees that it's acceptable for behavior to be undefined if users opt to muck with karpenter controlled annotations that aren't designed to be set by users.